### PR TITLE
Added to C API: adios2_version_struct. adios2_version(adios2_version_…

### LINKED ISSUE #3630: Want function to get ADIOS2 library version at run time
I would like to have a function (in the C API) that returns the ADIOS2 version number. This will allow the Julia wrapper to inquire the current version by calling the C API (or reading a global variable), and make run-time decisions based on this, e.g. whether certain functionality is available.

A C `#define` would not be good enough because these are not available in a shared library.

### PR DESCRIPTION
**Pull request by pnorbert**
…struct*) and adios2_version_str() functions.
Fixes #3630

### CODE REVIEW

**Comment by pnorbert** [comment]
@vicentebolea: On Linux I get this output mismatch since ADIOS2_VERSION_STR contains the tweak automatically generated by cmake. Can we have another string without it, or generate tweak on all platforms?
```
$ ./bin/hello_bpWriter_c
ADIOS version as struct 2.9.0
ADIOS version as string 2.9.0.12
```

File: bindings/C/adios2/c/adios2_c_types.h [hunk 1]
```diff
@@ -180,6 +180,13 @@ typedef struct
     adios2_blockinfo *BlocksInfo;
 } adios2_varinfo;
 
+typedef struct
+{
+    int major;
+    int minor;
+    int patch;
+} adios2_version_struct;
```
Comment (vicentebolea) [inline bindings/C/adios2/c/adios2_c_types.h:188]:
```suggestion
    int patch;
    char str[sizeof(ADIOS2_VERSION_STR)];
} adios2_version_struct;
```

File: bindings/C/adios2/c/adios2_c_adios.cpp [hunk 2]
```diff
@@ -17,6 +17,20 @@
 extern "C" {
 #endif
 
+char *adios2_version_str(void)
+{
+    char *str = (char *)malloc(sizeof(ADIOS2_VERSION_STR));
+    snprintf(str, sizeof(ADIOS2_VERSION_STR), ADIOS2_VERSION_STR);
+    return str;
+}
+
```
Comment (vicentebolea) [inline bindings/C/adios2/c/adios2_c_adios.cpp:26]:
```suggestion
```

File: bindings/C/adios2/c/adios2_c_adios.cpp [hunk 3]
```diff
@@ -17,6 +17,20 @@
...
+void adios2_version(adios2_version_struct *s)
+{
+    s->major = ADIOS2_VERSION_MAJOR;
+    s->minor = ADIOS2_VERSION_MINOR;
+    s->patch = ADIOS2_VERSION_PATCH;
+}
```
Comment (vicentebolea) [inline bindings/C/adios2/c/adios2_c_adios.cpp:32]:
```suggestion
    s->patch = ADIOS2_VERSION_PATCH;
    strncpy(s->str, ADIOS2_VERSION_STR, sizeof(ADIOS2_VERSION_STR));
}
```

File: examples/hello/bpWriter/helloBPWriter.c [hunk 4]
```diff
@@ -36,6 +36,14 @@ void check_handler(const void *handler, const char *message)
 
 int main(int argc, char *argv[])
 {
+    adios2_version_struct s;
+    adios2_version(&s);
+    printf("ADIOS version as struct %d.%d.%d\n", s.major, s.minor, s.patch);
+
+    char *vstr = adios2_version_str();
+    printf("ADIOS version as string %s\n", vstr);
+    free(vstr);
```
Comment (vicentebolea) [inline examples/hello/bpWriter/helloBPWriter.c:45]:
```suggestion
    printf("ADIOS version as string %s\n", s->str);
```

File: bindings/C/adios2/c/adios2_c_adios.h [hunk 5]
```diff
@@ -21,6 +21,20 @@
 extern "C" {
 #endif
 
+/**
+ * Return library version as string in major.minor.patch format
+ * String is allocated in the library, must be freed by caller.
+ */
+char *adios2_version_str(void);
+
```
Comment (vicentebolea) [inline bindings/C/adios2/c/adios2_c_adios.h:29]:
```suggestion
```

**Review by vicentebolea — COMMENTED** [review]
Looks good, I proposed a change to remove the dynamic alloc and reduce LoC. I do not feel too strong since this is a utility method, let me know :)

File: bindings/C/adios2/c/adios2_c_types.h (hunk 1 above)
Comment (eschnett) [reply bindings/C/adios2/c/adios2_c_types.h:188]: That will not work. This struct has a size that is not known in advance, and it is impossible to call it in a dynamic library. Almost every new version of ADIOS2 will have a different size for this struct, leading to ABI incompatibilities.

Why not use a `const char* str` field that points to a constant string that does not need to be deallocated?

File: bindings/C/adios2/c/adios2_c_types.h [hunk 6]
```diff
@@ -180,6 +180,13 @@ typedef struct
     adios2_blockinfo *BlocksInfo;
 } adios2_varinfo;
 
+typedef struct
```
Comment (eschnett) [inline bindings/C/adios2/c/adios2_c_types.h:183]: It is easier to call functions that don't expect or return a `struct`. I would prefer
```C
void adios2_version(int *major, int *minor, int *patch);
const char *adios2_version_str(void);
```
which are overall much simpler to use.

File: bindings/C/adios2/c/adios2_c_adios.cpp [hunk 7]
```diff
@@ -17,6 +17,20 @@
 extern "C" {
 #endif
 
+char *adios2_version_str(void)
```
Comment (eschnett) [inline bindings/C/adios2/c/adios2_c_adios.cpp:20]: Another option would be to use global variables, as in
```C
const int adios2_version_major;
const int adios2_version_minor;
const int adios_version_patch;
const char *const adios2_version_str;
```

**Comment by pnorbert** [comment]
Remove this pull request for a new one.

**Pull request closed without merging**